src/Numerics/SystemSolvers/batched_generalized_minimal_residual_solver.jl

```diff
@@ -1,167 +1,192 @@
+    batch_size::I
+    "total number of dofs per batched column"
+    dofperbatch::I
+    "residual norm in each column"
+    resnorms::resT
+    "initial residual norm in each column"
+    initial_resnorms::res0T
+    forward_reshape::FRS
+    forward_permute::FPR
+    backward_reshape::BRS
+    backward_permute::BPR
+
+    function BatchedGeneralizedMinimalResidual(
+        Q::AT,
+        dofperbatch,
+        Nbatch;
+        M = min(20, length(Q)),
+        rtol = √eps(eltype(AT)),
+        atol = eps(eltype(AT)),
+        forward_reshape = size(Q),
+        forward_permute = Tuple(1:length(size(Q))),
+    ) where {AT}
+        # Get ArrayType information
+        if isa(array_device(Q), CPU)
+            ArrayType = Array
+        else
+            # Sanity check since we don't support anything else
+            @assert isa(array_device(Q), CUDADevice)
+            ArrayType = CuArray
+        end
 
-# Return
-instance of `BatchedGeneralizedMinimalResidual` struct
-"""
-function BatchedGeneralizedMinimalResidual(
-    Qrhs;
-    m = size(Qrhs)[1],
-    n = size(Qrhs)[end],
-    subspace_size = m,
-    atol = sqrt(eps(eltype(Qrhs))),
-    rtol = sqrt(eps(eltype(Qrhs))),
-    ArrayType = Array,
-    reshape_tuple_f = size(Qrhs),
-    permute_tuple_f = Tuple(1:length(size(Qrhs))),
-)
-    k_n = subspace_size
-    # define the back permutations and reshapes
-    permute_tuple_b = Tuple(sortperm([permute_tuple_f...]))
-    tmp_reshape_tuple_b = [reshape_tuple_f...]
-    permute!(tmp_reshape_tuple_b, [permute_tuple_f...])
-    reshape_tuple_b = Tuple(tmp_reshape_tuple_b)
-    # allocate memory
-    residual = ArrayType(zeros(eltype(Qrhs), (k_n, n)))
-    b = ArrayType(zeros(eltype(Qrhs), (m, n)))
-    x = ArrayType(zeros(eltype(Qrhs), (m, n)))
-    sol = ArrayType(zeros(eltype(Qrhs), (m, n)))
-    rhs = ArrayType(zeros(eltype(Qrhs), (k_n + 1, n)))
-    cs = ArrayType(zeros(eltype(Qrhs), (2 * k_n, n)))
-    Q = ArrayType(zeros(eltype(Qrhs), (m, k_n + 1, n)))
-    H = ArrayType(zeros(eltype(Qrhs), (k_n + 1, n)))
-    R = ArrayType(zeros(eltype(Qrhs), (k_n + 1, k_n, n)))
-    return BatchedGeneralizedMinimalResidual(
-        atol,
-        rtol,
-        m,
-        n,
-        k_n,
-        residual,
-        b,
-        x,
-        sol,
-        rhs,
-        cs,
-        Q,
-        H,
-        R,
-        reshape_tuple_f,
-        permute_tuple_f,
-        reshape_tuple_b,
-        permute_tuple_b,
-    )
+        # FIXME: If we can batch the application of linearoperator!, then we dont
+        # need these two temporary work vectors (unpermuted/reshaped)
+        krylov_basis = similar(Q)
+        krylov_basis_prev = similar(Q)
+
+        FT = eltype(AT)
+        # Create storage for holding the batched Krylov basis
+        batched_krylov_basis =
+            fill!(ArrayType{FT}(undef, M + 1, dofperbatch, Nbatch), 0)
+
+        # Create storage for doing the batched Arnoldi process
+        Ω = fill!(ArrayType{FT}(undef, Nbatch, 2 * M), 0)
```

Comment:
Just a comment for the future: We may want to make the batched index the "slow index" in the future. In the simple tests I have run I haven't seen any issues with this though!

Reply:
Good point. I will track it in #1493 !